### ISSUE TITLE
Expose IPFS API as window.ipfs

### ISSUE THREAD

**Issue by lidel** [ipfs-shipyard/ipfs-companion#330]
There is an area of interest that can be summarized as.. 

> ..exposing a subset of IPFS APIs as `window.ipfs` on every webpage ✨ 

**tl;dr** If we have it, websites could detect that `window.ipfs` already exists and use it instead of spawning own `js-ipfs` node (saves resources, battery etc).

We had discussions in https://github.com/ipfs/js-ipfs-api/issues/387 and https://github.com/lidel/ipfs-firefox-addon/issues/37, https://github.com/ipfs/in-web-browsers/issues/9, https://github.com/ipfs/in-web-browsers/issues/35  but the consensus at the time was that it is not safe without some kind of access-control (eg. via proposed API Tokens: https://github.com/ipfs/go-ipfs/issues/1532).

@VictorBjelkholm  suggested a simple solution to the lack of access controls in the API itself: 
On the first use of `window.ipfs` user could be prompted with something like _"Do you want to allow scripts at `<website>` to access API of your IPFS node?"_. The choice would be remembered eg. for the page til the end of browsing session.  There could also be an additional _"remember for this site"_ checkbox to make UX better for users who wants to make permission persist between browser restarts. 

## Tasks

- [x] Create a [content script](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Anatomy_of_a_WebExtension) that creates a proxy object under `window.ipfs` of the page it is injected into
    - [x] This object should expose  a _subset_ of IPFS APIs
      > `<victorbjelkholm>`  there is of course a lot more once there is a full ipfs node in the browser, but if we only have the simplest and most useful functions (add, cat, pubsub subscribe/publish, dag get/put), it suddenly becomes a lot simpler and useful quickly
    - [x] Operations performed on `window.ipfs` should be proxied to the real IPFS API object present in webextension's [Background page](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Anatomy_of_a_WebExtension) (requires robust (de)serialization, we don't want to mangle upload data etc)
   - [x] First use of `window.ipfs` should trigger a dialog where user confirms that access to IPFS API is granted to requesting site (and decides if permission is for this session only or permanent)
- [x] Add a checkbox under "Preferences/Experiments" to control if content script  is injected to every visited page, make it disabled  by default initially
- [x] Write documentation, namely provide code snippet that enables apps to use the IPFS Companion IPFS node if it is available

### Potential problems

- [x] Something to investigate and solve before we make `window.ipfs` enabled by default: make sure it is not possible for a website to approve itself 🙃  This may mean we can't inject dialog to the tab itself and the dialog needs to be displayed  in a new Tab or via other means.

**Comment by VictorBjelkholm** [comment]
I've made two PoCs on this concept. One was separate and you can see it here: https://github.com/victorbjelkholm/js-ipfs-in-the-browser

The other one was a recent try to integrate that work into ipfs-companion and fix the stream issue in `.cat`, but didn't get really far yet: https://github.com/ipfs/ipfs-companion/commit/6b0ac8b6c0dbb45928a6e2ecaffed5d5e26f0ab7

>  suggested a simple solution to the lack of access controls in the API itself:
On the first use of window.ipfs user could be prompted with something like

I don't think that's good enough. I think (judging only five different methods) that the functions themselves should trigger the permission dialog (all with a "remember for this site" checkbox)

- `ipfs.id` would trigger a "website would like to know your ID" prompt
- `ipfs.cat` would trigger a "website would like to get content from the IPFS network via you or content in your local node" prompt
- `ipfs.pubsub.subscribe` would trigger a "website would like to read messages from a pubsub channel via you"
- `ipfs.pubsub.publish` would trigger a "website would like to send messages as you"

**Comment by VictorBjelkholm** [comment]
> make sure it is not possible for a website to approve itself
Should not be possible if we store it in the extensions storage.

>  This may mean we can't inject dialog to the tab itself and the dialog needs to be displayed in a new Tab or via other means.

We absolutely should **not** inject the dialog into the page itself, or via a new tab but rather use the popup that ipfs-companion provides. Take a look at how MetaMask deals with confirmation when you do a transaction.

**Comment by lidel** [comment]
@VictorBjelkholm thanks, great points!

For future reference, metamask webextension:
https://github.com/MetaMask/metamask-extension/tree/master/app

**Comment by diasdavid** [comment]
@lidel can a quick experiment to let users access window.ipfs after enabling a flag on the settings pane? @joaosantos15 would really appreciate this.

@joaosantos15, perhaps you can grab @VictorBjelkholm PoC and submit a PR to station?

**Comment by olizilla** [comment]
@diasdavid there is some very early wip for that here https://github.com/ipfs/ipfs-companion/pull/333

It's building on the ideas in https://github.com/victorbjelkholm/js-ipfs-in-the-browser
adding `window.ipfs` as a proxy to a ipfs instance running in ipfs-companion. Only `ipfs.id()` implemented so far... there are a lot of hurdles to jump through to make it work.

**Comment by lidel** [comment]
Just like @diasdavid suggested, we can work on PR #333 as an experiment that is disabled by default (another checkbox in  _Preferences_ 🙃 ). I feel that as long  it is an "opt-in experiment" we can ship it as a crude PoC, even without authorization GUI (although it may not be the hardest part, as noted in https://github.com/ipfs/ipfs-companion/pull/333#pullrequestreview-83146426).

**Comment by diasdavid** [comment]
Thanks @lidel @olizilla. Yes, shipping it as an opt-in would enable @joaosantos15 use case for now, that would help a ton :)

**Comment by olizilla** [comment]
@diasdavid  @joaosantos15 out of interest, what is the use case?

**Comment by diasdavid** [comment]
glad you asked :) it is for @joaosantos15 thesis -> https://github.com/ipfs/notes/issues/268 & https://github.com/inesc-id/hypercerts

**Comment by olizilla** [comment]
@diasdavid given all the limitations of proxy-ing across to an ipfs in the webext background page, would it be of any use if simply provide js-ipfs as a constructor at`window.IPFS` in the meantime? 

We can do both, it's just that a useable `window.ipfs` proxy is a little way off yet. So let pages construct their own jsipfs instance until we can proxy through to a shared one?

**Comment by joaosantos15** [comment]
Hello, sorry it has taken me this long to reply, and thank you for the help! :)
@olizilla, essentially what I need it to connect my web extension to the go-ipfs daemon (both running on the same machine). What you suggested,  
>  provide js-ipfs as a constructor atwindow.IPFS in the meantime?

would make that work, right?

Thanks!

**Comment by lidel** [comment]
@joaosantos15 ah, if you are developing a webextension, be sure to see our other proposal: 
["Provide API for other browser extensions"](https://github.com/ipfs/ipfs-companion/issues/307). Perhaps it is a better match?

Let us know which approach is better for you (`runtime.onMessageExternal` vs `window.ipfs`) and we can prioritize accordingly.

**Comment by joaosantos15** [comment]
Right now my extension is pretty simple, it only injects some js to change the layout of the webpage, I'd prefer `window.ipfs`.

**Comment by olizilla** [comment]
@joaosantos15 to add and get files with an ifps daemon you'll want `window.ipfs` set up as a proxy to an ipfs-api client hosted in the background page in ipfs-companion. The work in #333, when it's ready, will make that possible.

**Comment by VictorBjelkholm** [comment]
@joaosantos15 you could already today use `js-ipfs-api` easily to build what you need. The Ports or `window.ipfs` would just be a optimization to minimize dependencies, but what you need is already possible today.

If you check for `window.ipfs` or a ping message (interface to be defined) over a Port, before actually using it, once we have this feature it should work automagically.

Example:

```
if (window.ipfs) {
  // do whatever
} {
  // logic for loading js-ipfs-api, require/import async
  node = new ipfs()
  node.once('ready', () => { /* do whatever */ })
}
```

With this example, your application will work today and once `window.ipfs` or the Port is available, you'll automatically start using it.

**Comment by joaosantos15** [comment]
@VictorBjelkholm that's great. I'll go ahead and do it that way and when #333 is ready I'll switch to that. Thanks for the help, everyone 😄

**Comment by alanshaw** [comment]
Hey, some progress here:
https://youtu.be/t1ldUp_mjDk

* window.ipfs property!
    * (also adds window.Buffer if not there)
* proxies to running js-ipfs in the extension
    * video demonstrates shared repo across tabs
* Buffers sent across postMessage

WIP PR for companion here: https://github.com/ipfs/ipfs-companion/pull/333
IPFS `postMessage` proxy here: https://github.com/tableflip/ipfs-postmsg-proxy

**Comment by alanshaw** [comment]
Ok so, some more news.

As of v0.0.12, [ipfs-postmsg-proxy](https://github.com/tableflip/ipfs-postmsg-proxy) has feature parity with js-ipfs in the browser, according to the `interface-ipfs-core` tests, barring a [few small caveats](https://github.com/tableflip/ipfs-postmsg-proxy#caveats).

It's cool firstly because we're able to use `interface-ipfs-core` tests to ensure the proxy is working well and secondly because there's two types of tests:

1. Node.js tests - where we use mock window objects and `postMessage` APIs that use [`realistic-structured-clone`](https://www.npmjs.com/package/realistic-structured-clone) to simulate what happens in a browser when objects get passed between windows
2. Browser tests - because our Node.js simulation is good, but it's not the real thing and there _are_ differences between the two environments that we can't fake or mock. The browser tests setup a document with an iframe (where js-ipfs runs) to run the tests in an actual real life _real_ environment ftw

I've opened issues to deal with the caveats:
https://github.com/tableflip/ipfs-postmsg-proxy/issues/2
https://github.com/tableflip/ipfs-postmsg-proxy/issues/3
https://github.com/tableflip/ipfs-postmsg-proxy/issues/4

However my priority now is to implement https://github.com/tableflip/ipfs-postmsg-proxy/issues/5 which will allows us to add in the per method access control for ipfs companion.

I'll then do that in #333, and also implement a preference to enable/disable adding `ipfs` to the window

...and _then_, a first version can be shipped 🎉

**Comment by alanshaw** [comment]
Is this what we'd like for access control:

* Access control is fine grained, it requires permission to be granted to each IPFS function prior to usage, with option to allow access to all
* Granted permissions are scoped to the origin (site) on which they are granted and persist until the browser process exits. When granting permission, the user is given the option for the grant to be "remembered" for a given origin so that it persists across browser restarts
* Settings allow the user to revoke "remembered" privileges given to origins

**Comment by RangerMauve** [comment]
@alanshaw What is the definition of origin in this case?

**Comment by lidel** [comment]
Yes, we should have code in place to be able to control access at ("API function"+"Origin") level, but I feel we don't want to overwhelm end user with too many bells and whistles. 
I am quite sure UX may be tricky to get right here. 

What if we create a simplified access control to "allow all functions for current Origin" for now and take it for a spin with some PoC app?
It could provide us with some data points on how annoying access controls are :)

(we can experiment with access controls, as long this is an opt-in)

PS. when we say "Origin"  we think `window.location.origin` ([URL.origin](https://developer.mozilla.org/en-US/docs/Web/API/URL/origin)), right?

**Comment by RangerMauve** [comment]
I think that certain things should just be available by default
- block/dag/files/object API should just allow you to fetch and put stuff from IPFS, similar to how Android allows apps to access the network without any special permissions, and how you can do XHR in a browser without user consent
- DHT/swarm might need to be placed behind a permission since it has more low level access
- Pinning should be put behind a permission
- pubsub should be put behind a permission
- Once it's implemented fine grained permissions for encryption keys will be needed (per-key permissions)

From what I understand, getting a secure origin is still a WIP on devices that can't handle custom protocols or redirect to a proxy, so I'm not sure how useful permissions would be until that is addressed. 
Maybe it would make sense to do it based on which `multihash/path` is being used in the meantime, and later which `ipns/multihash` is being used in the future.

**Comment by lidel** [comment]
Indeed, Android-inspired model of allowing things like fetch by default may be a solution to my "UAC worries".  We should try do that as a PoC and see how it goes.

@RangerMauve  as for Origin, the issue you are thinking about is probably the [Problem #2](https://github.com/ipfs/in-web-browsers/blob/master/meeting-notes/2017-10-02--lab-week-discussions-on-ipfs-in-browser-extension.md#problem-2-inability-to-control-how-origin-is-calculated), but it is not really playing any relevant role here:  we want to expose `window.ipfs` on EVERY web page, not only on pages served by IPFS gateways. Gateways are a different problem space that can be tackled separately from this. The only side effect of mentioned problem is that user would have different origins while using the same IPFS path on two separate gateways – so an inconvenience, but not a  show-stopper. 
 
If we want to have access controls per website, we should just follow [standard browser security model based on Origin](https://en.wikipedia.org/wiki/Same-origin_policy).

**Comment by RangerMauve** [comment]
@lidel Thanks for clarifying about the origin issue.
The extension knows the address of the default gateway ahead of time, right?
What about having it detect whether the origin is the gateway, and applying access restrictions based on the hash that's being loaded?

For example,
- Assume the extension is configured use `https://ipfs.io/ipfs` as the gateway
- if a user navigates to https://ipfs.io/ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/wiki/InterPlanetary_File_System.html
- The plugin will detect that it's a URL for the gateway, and apply permissions for `ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco` instead of `https://ipfs.io`

**Comment by RangerMauve** [comment]
Alternately, auto-detect whether a gateway is being used by looking at the URL. If it matches `/ipfs/multihash*`, then take the origin, plus the multihash as the origin to use for the permissions. This way applications on any gateway are separated from each other, but they are also separated from different gateways so that a malicious gateway can't steal credentials from content served on a different gateway

**Comment by lidel** [comment]
I think until we have custom protocol handler capable of calculating Origin (https://github.com/ipfs-shipyard/ipfs-companion/issues/164) we should play it safe and create access control perimeter for  `window.ipfs`  based on: 
   - `window.location.origin` +  for a regular page  
   - `window.location.origin` + `/ipfs/$cid` if a page was loaded from IPFS path (gateway)

ps. Keep in mind that this discussion is only about access control for `window.ipfs` object (exposing IPFS API).
Things like cookies, localStorage etc are not covered by here, see https://github.com/ipfs-shipyard/ipfs-companion/issues/164

**Comment by alanshaw** [comment]
Progress on ACL for certain proxy actions:
https://youtu.be/m5ETDZ1HB24

**Comment by RangerMauve** [comment]
@alanshaw That's perfect! What permissions do you grant access to by default?

**Comment by alanshaw** [comment]
@RangerMauve currently these functions must be granted access https://github.com/tableflip/ipfs-companion/blob/e17bf70583229ef171a6c9af61bec60701263934/add-on/src/lib/ipfs-proxy/index.js#L21-L43

I made up that list so if you feel there's something in there that shouldn't be, or vice versa, then shout!

**Comment by RangerMauve** [comment]
That looks good to me! Only comment I'd have is that it would be useful to group permissions together.
For example, if something is going to be working with the object API, it would be useful to just have one popup instead of 6.

**Comment by alanshaw** [comment]
More progress today with access control and I think we're rapidly approaching a first cut version that can be merged:

Preferences option added:

<img width="1266" alt="screen shot 2018-01-12 at 16 02 59" src="https://user-images.githubusercontent.com/152863/34883732-5a8a53ac-f7b2-11e7-8d8e-d943f60cf0a1.png">

Permissions management page added:

<img width="1266" alt="screen shot 2018-01-12 at 16 02 38" src="https://user-images.githubusercontent.com/152863/34883759-6f8721ae-f7b2-11e7-9371-60db157443bf.png">

Check the video for more:
https://youtu.be/aXLERp9-Mis

**Comment by MirceaKitsune** [comment]
I am a Firefox user. I'm having a hard time understanding how to "teach" the browser about js-ipfs-api, in order for it to use IPFS specific JavaScript functions in website scripts. It would be very helpful if the IPFS Companion extension made this easier for the time being, by automatically exposing its interface to the browser! I'm glad to see there is progress on this issue, and would like to follow it further.

**Comment by diasdavid** [comment]
@kumavis check out the demo from @alanshaw on `postmsg-rpc` and exposing IPFS on window.ipfs https://youtu.be/qTOUMWujoEg . I believe you worked in something similar to add similar functionality to Metamask but that also worked for Service Workers, correct?

**Comment by lidel** [comment]
@alanshaw  to make sure we won't miss anything,  I've extracted remaining work into two issues:

- Isolate ACL for window.ipfs on IPFS paths (assume HTTP gateway) (#375)
- Custom dialog for window.ipfs ACLs (make it work under Firefox) (#377)

Do you feel there is anything else we should add as a dedicated issue?

**Comment by RangerMauve** [comment]
This is only for web pages and not service workers, correct? What should I do in order to get access to the local IPFS instance from a service worker?

**Comment by alanshaw** [comment]
Rough POC - https://github.com/ipfs/ipfs-service-worker/pull/7

**Comment by whyrusleeping** [comment]
Thinking through this a bit, on the permissions model, It would be really nice to be able to limit which directories in the files api that pages have access to. Maybe give them /apps/DOMAIN/ or something?

**Comment by whyrusleeping** [comment]
@alanshaw more like, restricting calls like `ipfs files cp`, `ipfs files read`, `ipfs files ls` etc to operating on paths under: `/apps/example.com/` or so.

**Comment by VictorBjelkholm** [comment]
We need to scope things based on not just the origin but the path as well. So an IPFS application would be scoped to `/apps/ipfs.io/ipfs/Qmasd/`. It's actually quite a nice model, and could also open up ways of applications sharing data via common directories, kind of how KBFS (from Keybase) works. `/keybase/private/victor` is mine, `/keybase/private/victor+why` is mine + why's.

**Comment by alanshaw** [comment]
I’m not sure I’m understanding you both properly...the permissions _are_ scoped to origin + pathname.

If you allow to /apps/example.com then anything under it will get access, but /apps/another.com will not.

**Comment by whyrusleeping** [comment]
@alanshaw I'm talking about paths within the ipfs files api.

**Comment by alanshaw** [comment]
Ah, got ya. I didn't pick up that you meant the actual paths of files stored in IPFS.

@VictorBjelkholm has suggested disabling all access to `files.*` until we can implement this. Which would allow us to ship this feature.

Thanks for taking the time to evaluate this @whyrusleeping - do you have any other concerns? I'm also wondering if there's more people that we should get to review this also...?

**Comment by alanshaw** [comment]
So just to flesh this out a little - we're talking about the [MFS specific functions of `ipfs.files`](https://github.com/ipfs/interface-ipfs-core/blob/master/SPEC/FILES.md#mutable-file-system).

When accessing MFS functions via `window.ipfs` your changes are limited to directories with the same scope as your application and only if the user grants permission.

e.g. if you application is running at https://example.com/mycoolapp then you can only access files and directories in or under `/https/example.com/mycoolapp`

To distinguish this scope from general access permissions scope we'll call it mfs-scope.

* `files.cp` - both source and destination path must be _below_ mfs-scope
* `files.mkdir` - path must be _below_ mfs-scope
* `files.stat` - path must be mfs-scope or below
* `files.rm` - path must be _below_ mfs-scope
* `files.read` - path must be _below_ mfs-scope
* `files.write` - path must be _below_ mfs-scope
* `files.mv` - both source and destination path must be _below_ mfs-scope
* `files.flush` - path must be _below_ mfs-scope
* `files.ls` - path must be mfs-scope or below

For the beta release I'll temporarily disable access to the above functions until I can implement the scope rules.

**Comment by alanshaw** [comment]
Is it worth namespacing access to companion? e.g. `/companion/https/example.com/mycoolapp`

**Comment by lidel** [comment]
I think it is a good idea, having `https` and `https` in root dir looks like a bad UX.
Perhaps `/dapps/https/example.com/mycoolapp` would be a better name? or just `apps-data` ?

**Comment by alanshaw** [comment]
I've been mulling this over and had a thought - should mfs-scope be transparent? So the app would read and write to `/*`, but that would be internally mapped onto `/app-data/https/example.com/*`?

I think that would be cool, but I don't know if there's any repercussions to doing this that I've perhaps not considered.

**Comment by lidel** [comment]
MFS is local and it should be fine for different apps to see different roots.
Transparent sandboxing of `ipfs.files` would be just a syntactic sugar, but makes developer's life more pleasant.

**Comment by alanshaw** [comment]
BREAKING NEWS ✨ PR is here ->> https://github.com/ipfs-shipyard/ipfs-companion/pull/431
Demo is here => https://youtu.be/KvgxAafphPg

**Comment by alanshaw** [comment]
@whyrusleeping @VictorBjelkholm something that came up today was sharing data between different versions of an application.

The MFS scoping does a good job of stopping other applications writing all over your stuff and each other's stuff but it means that if I have an app at ipfs.io/ipfs/Qm0 and I update it to ipfs.io/ipfs/Qm1 then ipfs.io/ipfs/Qm1 no longer has access to the files saved by ipfs.io/ipfs/Qm0, even though they are the same app.

It's been suggested that IPNS is the answer to this, but I wondered what your thoughts were?

We also discussed an alternative to MFS scoping - instead ask the user for permission to read/write to a directory (and sub directories).

**Comment by lidel** [comment]
Maybe this will help:

Let's take a look at Origin-based security context that isolates local storage, cookies etc in the regular web. Note that there is no GUI for opting out of Origin-based isolation per website. We have low-level CORS dance just so that one page can access API of the other.  But no GUI for escaping Origin sandbox. _That is for a good reason._

Here comes my main fear: as soon as we introduce UI for apps to ask user for access outside of its default security perimeter, we will end up with [Windows Vista UAC](https://en.wikipedia.org/wiki/User_Account_Control)-like hell of nagging dialogs.  Of course every app will want access to the root of your `ipfs.files`. Should it? IPFS introduces interesting space for innovation, as we can have zero-cost copies of the same data across multiple sandboxes.

I feel we should at the other direction: remove friction (dialogs) for every sandboxed method. Look closely at various methods and decide which ones are safe to be executed without ACL prompt.
Otherwise users will just learn to press "alllow" without reading, which defeats the purpose of ACLs.

**Comment by lidel** [comment]
### MFS Chroot/Sandboxing Just Landed

PR #431 just got merged. This means your app's MFS root directory is based on the origin and path where your application is running. 

Just for the record, I was playing with running this on various sites:

```js
ipfs.files.write('/test-from-' + window.location.host, Buffer.from('ehlo'), { create: true }).then(() => console.log('done'))
```
and sandboxing of `ipfs.files` works as expected :ok_hand: 

### Open Question: Merge IPFS Scopes?

It gets interesting when accessing `ipfs.files` via `window.ipfs` at IPNS site. 
I've run the snippet above at `/ipns/ipfs.io` and:
- when executed at Public Gateway `https://ipfs.io/ipns/ipfs.io/` 
  creates `/dapps/https/ipfs.io/ipns/ipfs.io/test-from-ipfs.io`
- when executed at Local Gateway `http://127.0.0.1:8080/ipns/ipfs.io/` 
  creates `/dapps/http/127.0.0.1%3A8080/ipns/ipfs.io/test-from-127.0.0.1:8080`

Do we want to merge IPFS scopes (`/ipns/` and `/ipfs/`) by removing HTTP component so that both above URLs will have access to the same `ipfs.files` root at `/dapps/ipns/ipfs.io/`  ?

**Comment by lidel** [comment]
Closing this issue: IPFS Companion [v2.2.0 shipped](https://blog.ipfs.io/35-ipfs-companion-2-2-0/) with `window.ipfs` enabled by default.


Remaining work related to refining security and UX constraints for window.ipfs  will be tracked in https://github.com/ipfs-shipyard/ipfs-companion/issues/454.

The [`window.ipfs` label](https://github.com/ipfs-shipyard/ipfs-companion/labels/window.ipfs) can be used for tracking all issues related to this endeavor.
